### PR TITLE
Keyboard Navigation in Coffee Script - Credit to @lowegreg 

### PR DESCRIPTION
**Pull request by Humni**
Credit to @lowegreg for the original js changes

### CODE REVIEW

**Comment by Humni** [comment]
Updated based on changes made to #109 - This version is untested and will need testing before merging

File: source/coffee/image-picker.coffee [hunk 1]
```diff
@@ -179,7 +189,46 @@ class ImagePickerOption
     imgAlt = @option.data("img-alt")
     if imgAlt
       image.attr('alt', imgAlt);
-    thumbnail.on("click", @clicked)  
+    thumbnail.on("click", @clicked)
```
Comment (rvera) [inline source/coffee/image-picker.coffee:207]: This is pretty big, can you make this another method and call it here?

File: source/coffee/image-picker.coffee [hunk 2]
```diff
@@ -179,7 +189,46 @@ class ImagePickerOption
...
+      exitingCtrl = !$(this).siblings().is($(event.relatedTarget).closest("li"));
+      $(this).closest("ul").find(".thumbnail").each () ->
+        if $(this).hasClass("selected") && exitingCtrl
+          $(this).prop("tabindex", "0");
+        else
+          $(this).prop("tabindex", "-1");
+        return
+      if $(this).closest("ul").find(".selected").length == 0 && exitingCtrl
+        $("li:first-child", $(this).closest("ul")).find(".thumbnail").prop("tabindex", "0");
+      return
```
Comment (rvera) [inline source/coffee/image-picker.coffee:219]: This looks pretty similar to the code above, should it be refactored?

File: source/sass/image-picker.scss [hunk 3]
```diff
@@ -22,25 +22,32 @@ ul.thumbnails.image_picker_selector
...
     {
       padding:  $thumbnail_padding;
       border:   $thumbnail_border;
+      border-radius: 4px;
       -webkit-user-select: none;
       -moz-user-select:    none;
       -ms-user-select:     none;
       img
       {
+        display:block;
```
Comment (rvera) [inline source/sass/image-picker.scss:49]: Why is this necessary?

File: source/sass/image-picker.scss [hunk 4]
```diff
@@ -22,25 +22,32 @@ ul.thumbnails.image_picker_selector
...
   {
-    margin:   0px $thumbnail_margin $thumbnail_margin 0px;
+    margin:   4px $thumbnail_margin $thumbnail_margin 4px;
     float:    left;
 
     .thumbnail
     {
       padding:  $thumbnail_padding;
       border:   $thumbnail_border;
+      border-radius: 4px;
```
Comment (rvera) [inline source/sass/image-picker.scss]: I'd prefer to keep this very minimal so each user styles this as needed. Is this change for any functionality or just because it looks better?

File: source/coffee/image-picker.coffee [hunk 5]
```diff
@@ -64,6 +64,16 @@ class ImagePicker
...
         option.unmark_as_selected()
+    jQuery("li", @picker).each () ->
+      if jQuery(this).hasClass("selected") && jQuery("li", @picker).find(":focus").length == 0
+        jQuery(this).prop("tabindex", "0");
+      else
+        jQuery(this).prop("tabindex", "-1");
+      return
+    if jQuery(this).find(":selected").length == 0 && jQuery("li", @picker).find(":focus").length == 0
+      $("li:first-child", @picker).find(".selectable").prop("tabindex", "0");
+    return
```
Comment (rvera) [inline source/coffee/image-picker.coffee:75]: Should these go 0, 1, 2, 3, etc? Or is it ok that they are all 0?

**Comment by rvera** [comment]
Left a couple of comments, otherwise lgtm!

File: source/coffee/image-picker.coffee (hunk 5 above)
Comment (Humni) [reply source/coffee/image-picker.coffee:75]: This is okay that they are all 0, means they are all 'tabbable' at the same priority, which then uses the order in the DOM to iterate through

File: source/sass/image-picker.scss (hunk 4 above)
Comment (Humni) [reply source/sass/image-picker.scss]: This can probably be removed

File: source/sass/image-picker.scss (hunk 3 above)
Comment (Humni) [reply source/sass/image-picker.scss:49]: Not sure, probably isn't needed. @lowegreg is this needed for any reason?

File: source/coffee/image-picker.coffee (hunk 2 above)
Comment (Humni) [reply source/coffee/image-picker.coffee:219]: Will look into refactoring this

File: source/coffee/image-picker.coffee (hunk 1 above)
Comment (Humni) [reply source/coffee/image-picker.coffee:207]: Sure thing

File: source/coffee/image-picker.coffee (hunk 5 above)
Comment (rvera) [reply source/coffee/image-picker.coffee:75]: Awesome, this is looking great!

File: source/sass/image-picker.scss (hunk 4 above)
Comment (lowegreg) [reply source/sass/image-picker.scss]: Border-radius isn't really necessary.  I did it originally to match the radius of the focus outline.

File: source/sass/image-picker.scss (hunk 3 above)
Comment (lowegreg) [reply source/sass/image-picker.scss:49]: The img display is a little more important.  If it's not block the image doesn't align vertically in the middle of the thumbnail div.  Leaving a larger blue selected area on the bottom.  Only a few pixels, but noticeable...

Comment (rvera) [reply source/sass/image-picker.scss:49]: Could this be `inline-block` then? so it behaves like a normal image.

Comment (lowegreg) [reply source/sass/image-picker.scss:49]: You could use inline-block if you adjust the line-height of the ul.thumbnails.image_picker_selector to 0px.  The img has an implicit line-height assigned to it. Since line-height only applies to inline elements, setting the img to display: block removes the line-height value.

File: source/sass/image-picker.scss (hunk 4 above)
Comment (lowegreg) [reply source/sass/image-picker.scss]: I should also mention that the addition of the 4px margin is needed in order to see the entire focus ring around an element.

**Comment by lowegreg** [comment]
Hold on guys, didn't realize that @Humni converted an older commit to coffee script.  The last commit I made was [0a11fbc](https://github.com/rvera/image-picker/pull/109/commits/0a11fbcedae4edee706fb82872e2418dadb05296). sorry for the confusion...

**Comment by Humni** [comment]
@lowegreg I have updated this pull request based on those 2 extra commits you made since my original pull request. See 3339151 commit message

File: source/sass/image-picker.scss (hunk 3 above)
Comment (Humni) [reply source/sass/image-picker.scss:49]: Any decision on which way to go with this?

File: source/sass/image-picker.scss (hunk 4 above)
Comment (Humni) [reply source/sass/image-picker.scss]: I say we should remove this for now, let the users modify as needed

**Comment by lowegreg** [comment]
Just checking back to see how this repo is going.  Did I forget to do something in order to have these requests merged???

G

**Comment by Humni** [comment]
@lowegreg I finally have push access, I'll update the pull request and let you know if it passes testing again

**Comment by Humni** [comment]
Testing passed:
 - Can tab left and right with arrow keys
 - Can tab left and right with tab and shift+tab
 - Can select with spacebar or enter/return key
